tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_socket.h"
+
+/* Used to override sendmsg and recvmsg for testing. */
+static ssize_t s2n_ktls_default_sendmsg(void *io_context, const struct msghdr *msg, uint8_t record_type);
+static ssize_t s2n_ktls_default_recvmsg(void *io_context, struct msghdr *msg, uint8_t *record_type);
+s2n_ktls_sendmsg_fn s2n_sendmsg_fn = s2n_ktls_default_sendmsg;
+s2n_ktls_recvmsg_fn s2n_recvmsg_fn = s2n_ktls_default_recvmsg;
+
+S2N_RESULT s2n_ktls_set_sendmsg_cb(struct s2n_connection *conn, s2n_ktls_sendmsg_fn send_cb, void *send_ctx)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(send_ctx);
+    RESULT_ENSURE(s2n_in_test(), S2N_ERR_NOT_IN_TEST);
+    conn->send_io_context = send_ctx;
+    s2n_sendmsg_fn = send_cb;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_set_recvmsg_cb(struct s2n_connection *conn, s2n_ktls_recvmsg_fn recv_cb, void *recv_ctx)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(recv_ctx);
+    RESULT_ENSURE(s2n_in_test(), S2N_ERR_NOT_IN_TEST);
+    conn->recv_io_context = recv_ctx;
+    s2n_recvmsg_fn = recv_cb;
+    return S2N_RESULT_OK;
+}
+
+/* TODO make use of record_type */
+static ssize_t s2n_ktls_default_recvmsg(void *io_context, struct msghdr *msg, uint8_t *record_type)
+{
+    POSIX_ENSURE_REF(io_context);
+    POSIX_ENSURE_REF(msg);
+    POSIX_ENSURE_REF(record_type);
+
+    int fd = 0;
+    const struct s2n_socket_read_io_context *peer_socket_ctx = io_context;
+    fd = peer_socket_ctx->fd;
+
+    return recvmsg(fd, msg, 0);
+}
```

Comment:
This still doesn't look quite right to me. You're not using `record_type`. Shouldn't the record type be parsed from `msg` by the caller? recvmsg only handles `msg`. If you put the logic for parsing `msg` in s2n_ktls_default_recvmsg, then you'll bypass it in all your unit tests.

You should only be mocking recvmsg, not any of your own logic. Same for sendmsg.

Reply:
This PR doesnt add a working definition of the default send/recvmsg; that comes in the following PRs. It only adds things necessary for mock stuffer impl.

Reply:
Yes, but shouldn't the only thing necessary for the mock be `io_context` and `msg`? `record_type` isn't relevant to recvmsg.

Reply:
technically `recvmsg` handles both data + ancillary. `recv` handles data only.

You are right, that its possible to write and parse the record type from `msghdr` but I can also mock that out and not deal with CMSG_* macro and allocation. I can also then test that functionality later in isolation.

Since this PR is large enough I am not inclined to add the writing/parsing CMSG logic to this PR. It makes more sense to add it in a following PR. Since this code is only used for testing, this should be ok.

Reply:
More importantly I think the reading/parsing CMSG deserves its own PR. Swapping this impl for the more complicated one will also serve as a good A/B test oracle.

Reply:
I agree that this mock interface doesn't seem correct. What's the purpose of passing the record type here? If it's just for the purpose of the mock impls that seems like the wrong path since it'll make the boundaries less clear. BTW you've already implemented setting/getting CMSG data since you have to do that anyway with the syscalls; just in reverse.

Reply:
I'll be handling this in a followup PR: https://github.com/aws/s2n-tls/pull/4109/files#diff-c9867e54a8d879980424cab3426b2d455215b438faedcaad673fbf33b1cd7aecR52

Reply:
Had some more discussions and I have added clarification for why this PR chooses to represent msg_control as `uint8_t` . https://github.com/aws/s2n-tls/pull/4109/commits/c051aae017dbf75bf5636cb9413a06b2134d6bfe

Here is a rough idea of what the next PR will look like:
It adds the `s2n_ktls_send` and `set_ancillary` methods and starts to represent the data on the send side as `cmsghdr`: https://github.com/aws/s2n-tls/compare/main...toidiu:s2n-tls:ak-ktls0_cmsg_send1?expand=1.

After that there will be a PR for the recv code path, thus making incremental progress towards what the final mock will look like.